permissions/resource_access_control_rule_set.go

```diff
@@ -0,0 +1,139 @@
+		}
+		workspaceClient, err := c.WorkspaceClient()
+		if err != nil {
+			return nil, err
+		}
+		return workspaceClient.AccountAccessControlProxy.GetRuleSet(ctx, getRuleSetReq)
+	}
+	updateThroughWsOrAcc := func(ctx context.Context, c *common.DatabricksClient, updateRuleSetReq iam.UpdateRuleSetRequest) (*iam.RuleSetResponse, error) {
+		if c.Config.AccountID != "" {
+			accountClient, err := c.AccountClient()
+			if err != nil {
+				return nil, err
+			}
+			return accountClient.AccessControl.UpdateRuleSet(ctx, updateRuleSetReq)
+		}
+		workspaceClient, err := c.WorkspaceClient()
+		if err != nil {
+			return nil, err
+		}
+		return workspaceClient.AccountAccessControlProxy.UpdateRuleSet(ctx, updateRuleSetReq)
+	}
+	fetchLatestEtagAndUpdateRuleSet := func(ctx context.Context, c *common.DatabricksClient, ruleSetUpdateReq iam.UpdateRuleSetRequest) (string, error) {
+		ruleSetGetRes, err := readFromWsOrAcc(ctx, c, iam.GetRuleSetRequest{
+			Name: ruleSetUpdateReq.Name,
+			Etag: "",
+		})
+		if err != nil {
+			return "", err
+		}
+		ruleSetUpdateReq.RuleSet.Etag = ruleSetGetRes.Etag
+		ruleSetUpdateRes, err := updateThroughWsOrAcc(ctx, c, ruleSetUpdateReq)
+		if err != nil {
+			return "", err
+		}
+		return ruleSetUpdateRes.Etag, nil
+	}
+	handleConflictAndUpdate := func(ctx context.Context, c *common.DatabricksClient, ruleSetUpdateReq iam.UpdateRuleSetRequest) (string, error) {
+		ruleSetUpdateRes, err := updateThroughWsOrAcc(ctx, c, ruleSetUpdateReq)
+		if err != nil {
+			var aerr *apierr.APIError
+			if !errors.As(err, &aerr) {
+				return "", err
+			}
+			if aerr.StatusCode == http.StatusConflict {
+				if aerr.ErrorCode == "RESOURCE_CONFLICT" {
+					// we need to get and update
+					etag, err := fetchLatestEtagAndUpdateRuleSet(ctx, c, ruleSetUpdateReq)
+					return etag, err
+				}
+			}
+			return "", err
+		}
+		return ruleSetUpdateRes.Etag, err
+	}
+	return common.Resource{
+		Schema: s,
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var ruleSetUpdateReq iam.UpdateRuleSetRequest
+			common.DataToStructPointer(d, s, &ruleSetUpdateReq.RuleSet)
+			ruleSetUpdateReq.Name = ruleSetUpdateReq.RuleSet.Name
+			etag, err := fetchLatestEtagAndUpdateRuleSet(ctx, c, ruleSetUpdateReq)
+			if err != nil {
+				return err
+			}
+			d.Set("etag", etag)
+			d.SetId(ruleSetUpdateReq.Name)
+			return nil
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			data, err := readFromWsOrAcc(ctx, c, iam.GetRuleSetRequest{
+				Name: d.Id(),
+				Etag: d.Get("etag").(string),
+			})
+			if err != nil {
+				return err
+			}
+			return common.StructToData(data, s, d)
+		},
+		Update: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			var ruleSetUpdateReq iam.UpdateRuleSetRequest
+			common.DataToStructPointer(d, s, &ruleSetUpdateReq.RuleSet)
+			ruleSetUpdateReq.Name = ruleSetUpdateReq.RuleSet.Name
```

Comment:
Not a blocker for this PR, but this is a bit unusual for an API. I think the API you're exposing in Terraform makes sense, as does the Read API, but why does Name need to be duplicated to the top-level structure for update and create? Why does there even need to be a wrapper structure at all? From my reading, your API should be able to accept a RuleSet directly in the update call since RuleSet contains its own name. In this case, the request/response body structure will be identical for create, read, and update.

Reply:
This is good observation Miles. 

> but why does Name need to be duplicated to the top-level structure for update and create?

The top level name is used to reference the rule set that is being updated. And the wrapper structure is used to prevent customers from blinding passing the response from `Get` to `Update`. In our workspace permissions API, we have encountered situations where customers just passed the response from GET which included inherited permissions to PUT and caused a permission explosion. So we have intentionally tried to make sure the same mistake cannot happen in the rule set API. 

@jiyunyao can you also please take a look here?

Reply:
I can see that happening since this API is PUT (i.e. replace), not PATCH (i.e. incremental change). I worry that changes like this will be harder to support in the future when we try to automate the generation of Terraform resource implementations, and SDK users also may be confused about what name parameter to use for https://github.com/databricks/databricks-sdk-go/blob/main/service/iam/model.go#L634 and for https://github.com/databricks/databricks-sdk-go/blob/main/service/iam/model.go#L610.

Reply:
We are following AIP: https://google.aip.dev/134#etags where the resource body contains the name and update also has the name in URL: https://google.aip.dev/134#guidance.

This is similar design as https://cloud.google.com/iam/docs/reference/rest/v2/policies/update.

We don't intend to introduce PATCH for RuleSet update as it doesn't make sense (you can't partial update a list easily). We may support patch on individual rule update.

Reply:
@mgyucht I remember now. The name is part of the body because [App Infra](https://databricks.atlassian.net/browse/APP-1109) still doesn't support path parameters containing slash as recommended by the AIP @jiyunyao linked above.